lib/net.c

```diff
@@ -219,6 +233,120 @@ ipv4_ntuple_filter_add(struct gatekeeper_if *iface, rte_be32_t dst_ip_be,
 	return 0;
 }
 
+static int
+ipv4_flow_add(struct gatekeeper_if *iface, rte_be32_t dst_ip_be,
+	rte_be16_t src_port_be, rte_be16_t src_port_mask_be,
+	rte_be16_t dst_port_be, rte_be16_t dst_port_mask_be,
+	uint8_t proto, uint16_t queue_id)
+{
+	struct rte_flow_attr attr = { .ingress = 1, };
+	struct rte_flow_action_queue queue = { .index = queue_id };
+	struct rte_flow_action action[2] = {
+		{
+			.type = RTE_FLOW_ACTION_TYPE_QUEUE,
+			.conf = &queue,
+	       	},
+		{
+			.type = RTE_FLOW_ACTION_TYPE_END,
+		}
+	};
+	struct rte_flow_item pattern[3] = {
+		{
+			.type = RTE_FLOW_ITEM_TYPE_ETH,
+			.spec = {
+				.type = rte_cpu_to_be_16(RTE_ETHER_TYPE_IPV4),
+			},
+			.mask = {
+				.type = 0xFFFF,
+			},
+		},
+		{
+			.type = RTE_FLOW_ITEM_TYPE_IPV4,
+			.spec = {
+				.hdr = {
+					.dst_addr = dst_ip_be,
+					.next_proto_id = proto,
+				}
+			},
+			.mask = {
+				.hdr = {
+					.dst_addr = 0xFFFFFFFF,
+					.next_proto_id = 0xFF,
+				}
+			},
+		},
+		{
+			.type = RTE_FLOW_ITEM_TYPE_END,
+		},
+	};
+	struct rte_flow *flow;
+	struct rte_flow_item_tcp tcp_spec;
+	struct rte_flow_item_tcp tcp_mask;
+	struct rte_flow_item_udp udp_spec;
+	struct rte_flow_item_udp udp_mask;
+	struct rte_flow_error error;
+	int ret;
+
+	if (!iface->rss) {
+		/*
+		 * IPv4 flows can only be used if supported by the NIC
+		 * (to steer matching packets) and if RSS is supported
+		 * (to steer non-matching packets elsewhere).
+		 */
+		G_LOG(NOTICE, "net: cannot use IPv4 flows when RSS is not supported\n");
+		return -1;
+	}
+
+	if (proto == IPPROTO_TCP) {
+		memset(&tcp_spec, 0, sizeof(struct rte_flow_item_tcp));
+		memset(&tcp_mask, 0, sizeof(struct rte_flow_item_tcp));
```

Comment:
Replace `sizeof(struct rte_flow_item_tcp)` with `sizeof(tcp_spec)` and `sizeof(tcp_mask)`, accordingly.